extra/camino/man/man1/subsetscheme.1
.\" $Id$

.TH subsetscheme 1

.SH NAME
subsetscheme \- Create scheme files from a subset of directions in an existing scheme

.SH SYNOPSIS
.B subsetscheme -schemefile \fR <\fIfile\fR> \fB-subsetpoints\fR <\fIpoints\fR> 
\fB-outputroot\fR <\fIroot\fR> [\fB-imagelist\fR <\fIlist\fB>]

 
.SH DESCRIPTION

This program takes the output of \fBsubsetpoints\fR and produces scheme files
corresponding to the subsets. Each subset scheme file contains ALL the unweighted (q=0)
measurements in the original scheme, as well as all measurements along the gradient
directions in the subset.


.SH OUTPUT

Subset scheme files - the measurements in the original scheme that correspond to the
gradient directions in the subsetpoints file are output as outputRoot_[0...N-1].scheme,
where there are N subsets.

Subset indices - the integer indices of the measurements from the full scheme that are in
the subset, indexed from zero.

Image lists - If an image list is provided with the \fB-imagelist\fR option, the subset
image lists are produced. The image list should be one line per measurement, containing the 
file name of the associated image volume. This can be produced from a 4D Nifti image with 
\fBsplit4dnii\fR.


.SH EXAMPLES

Begin with an existing scheme file, A.scheme, which consists of 3 b=0 measurements and 60
b=1000 s / mm^2 measurements spread over the sphere. The 60 diffusion-weighted
measurements are stored in the file A.points.

   subsetpoints -inputfile A.points -pointspersubset 20 20 20 -outputfile A_subsets

Create scheme files for each of the subsets

   subsetscheme -schemefile A.scheme -subsetpoints A_subsets -outputroot subset

The files subset_[0,1,2].scheme are created, each containing the 3 b=0 measurements,
followed by the b=1000 measurements in each subset.

If the diffusion-weighted images are stored as Nifti volumes, an image list can also be
created for the subsets. If all 63 volumes are listed in the file imagelist.txt, then the
command

    subsetscheme -schemefile A.scheme -subsetpoints A_subsets -outputroot subset -imagelist imagelist.txt

will also produce subset_[0,1,2].imagelist, which contains the volumes in each subset.


.SH OPTIONS

.TP
.B \-schemefile\fR 
Scheme file for the complete imaging scheme. 

.TP
.B \-subsetpoints\fR 
Output from subsetpoints containing one or more subsets of the directions in the scheme
file.

.TP
.B \-outputroot\fR 
Root for the output file names.

.TP
.B \-imagelist\fR 
Image list containing file names of raw DWI images for the complete imaging scheme. These
will  be parsed and an image list will be created for each subset.

.SH "AUTHORS"
Philip Cook <camino@cs.ucl.ac.uk>

.SH "SEE ALSO"
pointset2scheme(1), subsetpoints(1)

.SH BUGS


extra/camino/apps/SubsetScheme.java
package apps;

import imaging.*;
import misc.*;
import numerics.*;
import tools.*;
import data.*; //added for OutputManager

import java.io.*;
import java.util.logging.*;
import java.util.Random;
import java.util.Scanner;
import java.text.*;


/**
 * Produces information about the quality of electrostatic subsets, and produces scheme
 * files for those subsets given the scheme file for the full set.
 *
 * @author Philip Cook
 * @version $Id$
 */
public class SubsetScheme extends Executable{
    
    /**
     * logging object
     */
    private static final Logger logger = Logger.getLogger("apps/SubsetScheme");
    
    
    public SubsetScheme(String[] args){
        super(args);
    }
    
    // contains non-zero normalized qs
    int fullSetSize;
    double[][] fullSetPoints;
    int numSubsets;
    // non zero qs in subsets. Subset schemes will contain these plus
    // all q=0 measurements
    int[] subsetSizes;
    // subsetPoints[s] is the set of points for set s
    double[][][] subsetPoints;
    // total measurements in full scheme
    int fullMeas;
    String imageListFile;
    String[] imageList;
    String outputRoot;
    String subsetFile;
    
    public void initDefaultVals(){
        fullSetSize = 0;
        fullSetPoints = null;
        numSubsets = 0;
        subsetSizes = null;
        subsetPoints = null;
        fullMeas = 0;
        imageListFile = null;
        imageList = null;
        outputRoot = "subset";
        subsetFile = null;
    }
    
    public void initOptions(String[] args){
        //    public static void main(String[] args) {
        
        if (args.length == 0) {
            System.err.println("SubsetScheme -schemefile <schemefile> -subsetpoints <points> [-imagelist <list> -outputroot <root>]");
            System.exit(0);
        }
        
        CL_Initializer.CL_init(args);
        
        CL_Initializer.initImagingScheme();
        
        for (int i = 0; i < args.length; i++) {
            if (args[i].equals("-subsetpoints")) {
                subsetFile = args[i+1];
                CL_Initializer.markAsParsed(i,2);
            }
            if (args[i].equals("-imagelist")) {
                imageListFile = args[i+1];
                CL_Initializer.markAsParsed(i,2);
            }
            if (args[i].equals("-outputroot")) {
                outputRoot = args[i+1];
                CL_Initializer.markAsParsed(i,2);
            }
        }
        
        CL_Initializer.checkParsing(args);
    }
    
    public void initVariables(){
        fullMeas = CL_Initializer.imPars.numMeasurements();
        
        fullSetPoints = CL_Initializer.imPars.getNonZeroG_Dirs();
        
        fullSetSize = fullSetPoints.length;
    }
    
    
    public void execute(OutputManager om){
        // get subset points
        try {
            Scanner subsetHeader = new Scanner(new File(subsetFile)).useDelimiter("\\s+");
            
            // total number of points in subset scheme
            int totalSubsetPoints = subsetHeader.nextInt();
            
            if (fullSetSize != totalSubsetPoints) {
                throw new LoggedException("Full scheme has " + fullSetSize + " non-zero g, but " +
                "subset is from a set of " + totalSubsetPoints + " points");
            }
            
            numSubsets = subsetHeader.nextInt();
            
            subsetSizes = new int[numSubsets];
            
            for (int s = 0; s < numSubsets; s++) {
                subsetSizes[s] = subsetHeader.nextInt();
            }
            
            // file contains all points in scheme
            double[][] tmp = PointSetToScheme.readPoints(subsetFile, false, false, false);
            
            int pointCounter = 0;
            
            subsetPoints = new double[numSubsets][][];
            
            for (int s = 0; s < numSubsets; s++) {
                
                subsetPoints[s] = new double[subsetSizes[s]][];
                
                for (int p = 0; p < subsetSizes[s]; p++) {
                    subsetPoints[s][p] = tmp[pointCounter++];
                }
                
            }
            
            
            
        }
        catch (IOException e) {
            throw new LoggedException(e);
        }
        
        
        
        
        if (imageListFile != null) {
            try {
                Scanner images = new Scanner(new File(imageListFile));
                
                images.useDelimiter("\\n");
                
                imageList = new String[fullMeas];
                
                for (int i = 0; i < fullMeas; i++) {
                    imageList[i] = images.next();
                }
                
            }
            catch (IOException e) {
                logger.warning("Can't process image list " + imageListFile + ", image list will " +
                "not be written");
                
                imageList = null;
            }
        }
        
        for (int s = 0; s < numSubsets; s++) {
            
            int numB0 = CL_Initializer.imPars.numZeroMeasurements();
            
            int subsetMeas = subsetSizes[s] + numB0;
            
            // which measurements are in the subset
            int[] subsetIndices = new int[subsetMeas];
            
            int indexCounter = 0;
            
            // iterate over measurements in scheme, not points in fullSet (want to get
            // the indices right including the b=0).
            
            // first get all of the b=0 measurements
            for (int m = 0; m < fullMeas; m++) {
                if (CL_Initializer.imPars.zero(m)) {
                    subsetIndices[indexCounter++] = m;
                }
                
            }
            
            for (int i = 0; i < subsetSizes[s]; i++) {
                
                // subset measurement is the closest measurement to the subset direction
                
                int maxInd = 0;
                
                double maxDot = 0.0;
                
                for (int m = 0; m < fullMeas; m++) {
                    
                    double[] g = CL_Initializer.imPars.getG_Dir(m);
                    
                    // no Math.abs here because we are looking for a subset of the directions,
                    // not a common axis
                    // precision errors prevent an exact match
                    double dotProd =
                    g[0] * subsetPoints[s][i][0] +
                    g[1] * subsetPoints[s][i][1] +
                    g[2] * subsetPoints[s][i][2];
                    
                    // fails if we can't resolve all points, but this should not happen with
                    // a decent angular separation (> 1 degree should be fine)
                    if (dotProd > maxDot) {
                        maxInd = m;
                        maxDot = dotProd;
                    }
                }
                
                
                // sanity check
                if (maxDot < 0.99984769516) {
                    logger.warning("Couldn't match point " + i + " in subset " + s +
                    " to a scheme direction within 1 degree. Closest angle was " +
                    Math.acos(maxDot) * 180.0 / Math.PI);
                }
                
                subsetIndices[indexCounter++] = maxInd;
                
            }
            
            if (indexCounter != subsetIndices.length) {
                throw new
                LoggedException("Could not match all points in subset to points " +
                "in the full set");
            }
            
            
            DW_Scheme subsetScheme = CL_Initializer.imPars.getSubsetScheme(subsetIndices);
            FileOutput out = new FileOutput(outputRoot +  "_" + s + ".scheme");
            out.writeString(subsetScheme.toString());
            out.close();
            
            out = new FileOutput(outputRoot +  "_" + s + ".indices");
            
            for (int i = 0; i < subsetMeas; i++) {
                out.writeString(subsetIndices[i] + "\n");
            }
            
            out.close();
            
            if (imageList != null) {
                
                out = new FileOutput(outputRoot + "_" + s + ".imagelist");
                
                for (int i = 0; i < subsetMeas; i++) {
                    out.writeString(imageList[subsetIndices[i]] + "\n");
                }
                
                out.close();
            }
            
            // now test how this subset ranks against a randomly chosen sample
            
            DecimalFormat df = new DecimalFormat("0.000");
            
            double[][] optimalSet = SphericalPoints.getElecPointSet(subsetSizes[s]);
            
            double[] randomEnergy = getRandomSubsetEnergy(fullSetPoints, subsetSizes[s], 100,
            new MTRandom(CL_Initializer.seed));
            
            double meanRandomEnergy = ArrayOps.mean(randomEnergy);
            double medianRandomEnergy = ArrayOps.median(randomEnergy);
            
            System.out.println("Energy of subset " + s + "\t\t" + df.format(pointsetEnergy(subsetPoints[s])));
            System.out.println("Energy of optimal " + subsetSizes[s] + "-point set\t" +
            df.format(pointsetEnergy(optimalSet)));
            System.out.println("Mean energy of random subset\t" +
            df.format(meanRandomEnergy));
            System.out.println("Median energy of random subset\t" +
            df.format(medianRandomEnergy));
            System.out.println();
        }
        
    }
    
    
    
    /**
     * Gets a randomly chosen subset of points.
     *
     *
     */
    public static double[][] getRandomSubset(double[][] points, int size, Random ran) {
        
        int fullSize = points.length;
        
        boolean[] chosen = new boolean[fullSize];
        
        int numChosen = 0;
        
        while (numChosen < size) {
            
            int next = ran.nextInt(fullSize);
            
            if (!chosen[next]) {
                chosen[next] = true;
                numChosen++;
            }
        }
        
        int indexCounter = 0;
        
        double[][] subsetPoints = new double[size][3];
        
        for (int i = 0; i < fullSize; i++) {
            if (chosen[i]) {
                subsetPoints[indexCounter][0] = points[i][0];
                subsetPoints[indexCounter][1] = points[i][1];
                subsetPoints[indexCounter][2] = points[i][2];
                indexCounter++;
            }
        }
        
        return subsetPoints;
        
    }
    
    
    /**
     * Gets a sample of random subsets, and returns their energy.
     *
     *
     */
    public static double[] getRandomSubsetEnergy(double[][] points, int size, int samples,
    Random ran) {
        
        double[] subsetEnergy = new double[samples];
        
        for (int i = 0; i < samples; i++) {
            double[][] subset = getRandomSubset(points, size, ran);
            subsetEnergy[i] = pointsetEnergy(subset);
        }
        
        return subsetEnergy;
        
    }
    
    
    /**
     * Calculates the total energy of a point set.
     *
     */
    public static double pointsetEnergy(double[][] points) {
        double[][] energy = OrderedAcqMinimizer.pairEnergyMatrix(points);
        
        double subsetEnergy = 0.0;
        
        for (int n = 0; n < points.length; n++) {
            for (int m = 0; m < n; m++) {
                subsetEnergy += energy[n][m];
            }
        }
        
        return subsetEnergy;
    }
    
}
